homeassistant/components/sensor/snmp.py

```diff
@@ -66,27 +105,69 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
     baseoid = config.get(CONF_BASEOID)
     unit = config.get(CONF_UNIT_OF_MEASUREMENT)
     version = config.get(CONF_VERSION)
+    username = config.get(CONF_USERNAME)
+    authkey = config.get(CONF_AUTH_KEY)
+    authproto = config.get(CONF_AUTH_PROTOCOL)
+    privkey = config.get(CONF_PRIV_KEY)
+    privproto = config.get(CONF_PRIV_PROTOCOL)
     accept_errors = config.get(CONF_ACCEPT_ERRORS)
     default_value = config.get(CONF_DEFAULT_VALUE)
     value_template = config.get(CONF_VALUE_TEMPLATE)
 
     if value_template is not None:
         value_template.hass = hass
 
-    errindication, _, _, _ = next(
-        getCmd(SnmpEngine(),
-               CommunityData(community, mpModel=SNMP_VERSIONS[version]),
-               UdpTransportTarget((host, port)),
-               ContextData(),
-               ObjectType(ObjectIdentity(baseoid))))
+    if version == '3':
+        auth_protocols = {
+            'none': usmNoAuthProtocol,
+            'hmac-md5': usmHMACMD5AuthProtocol,
+            'hmac-sha': usmHMACSHAAuthProtocol,
+            'hmac128-sha224': usmHMAC128SHA224AuthProtocol,
+            'hmac192-sha256': usmHMAC192SHA256AuthProtocol,
+            'hmac256-sha384': usmHMAC256SHA384AuthProtocol,
+            'hmac384-sha512': usmHMAC384SHA512AuthProtocol,
+        }
+        priv_protocols = {
+            'none': usmNoPrivProtocol,
+            'des': usmDESPrivProtocol,
+            '3des-ede': usm3DESEDEPrivProtocol,
+            'aes-cfb-128': usmAesCfb128Protocol,
+            'aes-cfb-192': usmAesCfb192Protocol,
+            'aes-cfb-256': usmAesCfb256Protocol,
+        }
+
+        request_args = [
+            SnmpEngine(),
+            UsmUserData(
+                username,
+                authKey=authkey or None,
+                privKey=privkey or None,
```

Comment:
Need not set to None

Reply:
This replaces en empty string with None, which makes a difference in UsmUserData. Are you implying that authkey=="" is guaranteed not to appear in this codepath?